### PR TITLE
`CGOError` --> `CGOErrCode`

### PR DESCRIPTION
**Pull request by ibeckermayer**
Rather than passing errors via C-strings, we pass them as integers which removes a major potential memory leak/error footgun

(Credit to @zmb3 for this idea)

### CODE REVIEW

**Review by strideynet — APPROVED** [review]
I'm not overly familiar with Rust, but the Go/CGo changes look sensible.

Am I right in thinking that with this change we lose some detail on the Go side as to what error actually occurred, but that these errors are now logged in the Rust side ?

File: lib/srv/desktop/rdp/rdpclient/src/lib.rs [hunk 1]
```diff
@@ -677,33 +698,18 @@ unsafe fn from_go_array(len: u32, ptr: *mut u8) -> Vec<u8> {
...
-///
-/// The pointer inside the CGOError must point to a valid null terminated Go string.
-unsafe fn from_cgo_error(e: CGOError) -> String {
-    let s = from_go_string(e);
-    free_go_string(e);
-    s
+#[repr(C)]
+#[derive(Copy, Clone, PartialEq)]
+pub enum CGOErrCode {
+    ErrCodeSuccess,
```
Comment (zmb3) [inline lib/srv/desktop/rdp/rdpclient/src/lib.rs:704]: Can you add numeric values here to ensure the `ErrCodeSuccess` gets the value 0, which is traditional for C-based APIs to indicate no error.

**Comment by zmb3** [comment]
> Am I right in thinking that with this change we lose some detail on the Go side as to what error actually occurred, but that these errors are now logged in the Rust side ?

It's actually the opposite. We were passing a string from Go to Rust, just so Rust could log the error and then terminate.

Instead, we log the error on the Go side, and then just indicate to Rust that there was a failure. It's a lot easier to pass an integer than a string across the FFI.

File: lib/srv/desktop/rdp/rdpclient/src/lib.rs (hunk 1 above)
Comment (ibeckermayer) [reply lib/srv/desktop/rdp/rdpclient/src/lib.rs:704]: I believe `#[repr(C)]` gives us that here per C's default behavior, no harm making it explicit though.

File: lib/srv/desktop/rdp/rdpclient/librdprs.h [hunk 2]
```diff
@@ -46,14 +51,9 @@ typedef enum CGOPointerWheel {
...
 
-/**
- * CGOError is an alias for a C string pointer, for C API clarity.
- */
-typedef char *CGOError;
-
 typedef struct ClientOrError {
   struct Client *client;
-  CGOError err;
+  enum CGOErrCode err;
```
Comment (nklaassen) [inline lib/srv/desktop/rdp/rdpclient/librdprs.h:56]: If I'm remembering my C - Above, you typedef'd `enum CGOErrCode` to just `CGOErrCode`, so I don't think you need `enum` before it everywhere unless this is necessary for the Rust side

File: lib/srv/desktop/rdp/rdpclient/client.go [hunk 3]
```diff
@@ -283,16 +282,16 @@ func (c *Client) start() {
...
 					C.CGOMousePointerEvent{
 						x:      C.uint16_t(m.X),
 						y:      C.uint16_t(m.Y),
 						button: C.PointerButtonNone,
 						wheel:  C.PointerWheelNone,
 					},
-				)); err != nil {
-					c.cfg.Log.Warningf("Failed forwarding RDP mouse pointer: %v", err)
+				); err != C.ErrCodeSuccess {
+					c.cfg.Log.Warningf("write_rdp_pointer failed on the Rust side")
```
Comment (nklaassen) [inline lib/srv/desktop/rdp/rdpclient/client.go:294]: The new error message seems less helpful to a user, not sure they care about "on the Rust side" "on the Go side", why not leave "Failed forwarding RDP mouse pointer" here?

File: lib/srv/desktop/rdp/rdpclient/librdprs.h (hunk 2 above)
Comment (zmb3) [reply lib/srv/desktop/rdp/rdpclient/librdprs.h:56]: This header is autogenerated by `cbindgen`.

File: lib/srv/desktop/rdp/rdpclient/client.go (hunk 3 above)
Comment (ibeckermayer) [reply lib/srv/desktop/rdp/rdpclient/client.go:294]: [That error message has now been moved over to Rust](https://github.com/gravitational/teleport/blob/2cb663db3cda686e3e5be243ba7d94f3695e9d66/lib/srv/desktop/rdp/rdpclient/src/lib.rs#L593). Agreed that this isn't important to the user though, I will change all of these to `Debugf`'s.

Comment (zmb3) [reply lib/srv/desktop/rdp/rdpclient/client.go]: I still think there's no need to log "on the Rust side" - this is kind of irrelevant and will just confuse a user reading their logs.

**Review by nklaassen — APPROVED** [review]

**Review by zmb3 — APPROVED** [review]

**Pull request merged**